llvm/include/llvm/Telemetry/Telemetry.h

```diff
@@ -0,0 +1,136 @@
+///              telemery data.
+/// - TelemetryInfo: data courier
+/// - TelemetryConfig: this stores configurations on Telemeter.
+///
+/// Refer to its documentation at llvm/docs/Telemetry.rst for more details.
+//===---------------------------------------------------------------------===//
+
+#ifndef LLVM_TELEMETRY_TELEMETRY_H
+#define LLVM_TELEMETRY_TELEMETRY_H
+
+#include <chrono>
+#include <ctime>
+#include <memory>
+#include <optional>
+#include <string>
+
+#include "llvm/ADT/StringExtras.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/Error.h"
+#include "llvm/Support/JSON.h"
+
+namespace llvm {
+namespace telemetry {
+
+/// Configuration for the Telemeter class.
+/// This stores configurations from both users and vendors and is passed
+/// to the Telemeter upon construction. (Any changes to the config after
+/// the Telemeter's construction will not have effect on it).
+///
+/// This struct can be extended as needed to add additional configuration
+/// points specific to a vendor's implementation.
+struct Config {
+  // If true, telemetry will be enabled.
+  bool EnableTelemetry;
+
+  // Implementation-defined names of additional destinations to send
+  // telemetry data (Could be stdout, stderr, or some local paths, etc).
+  //
+  // These strings will be interpreted by the vendor's code.
+  // So the users must pick the from their vendor's pre-defined
+  // set of Destinations.
+  std::vector<std::string> AdditionalDestinations;
+};
+
+/// For isa, dyn_cast, etc operations on TelemetryInfo.
+typedef unsigned KindType;
+/// This struct is used by TelemetryInfo to support isa<>, dyn_cast<>
+/// operations.
+/// It is defined as a struct (rather than an enum) because it is
+/// expectend to be extended by subclasses which may have
+/// additional TelemetryInfo types defined to describe different events.
+struct EntryKind {
+  static const KindType Base = 0;
+};
+
+/// TelemetryInfo is the data courier, used to move instrumented data
+/// from the tool being monitored to the Telemery framework.
+///
+/// This base class contains only the basic set of telemetry data.
+/// Downstream implementations can add more fields as needed to describe
+/// additional events.
+///
+/// For example, The LLDB debugger can define a DebugCommandInfo subclass
+/// which has additional fields about the debug-command being instrumented,
+/// such as `CommandArguments` or `CommandName`.
+struct TelemetryInfo {
+  // This represents a unique-id, conventionally corresponding to
+  // a tool's session - i.e., every time the tool starts until it exits.
+  //
+  // Note: a tool could have multiple sessions running at once, in which
+  // case, these shall be multiple sets of TelemetryInfo with multiple unique
+  // ids.
+  //
+  // Different usages can assign different types of IDs to this field.
+  std::string SessionId;
+
+  TelemetryInfo() = default;
+  virtual ~TelemetryInfo() = default;
+
+  virtual json::Object serializeToJson() const;
```

Comment:
I'm wondering if we've got slightly different visions of things and it's driving the way we're looking at the design differently. Reading between the lines of some of the comments and responses you've made, I get the impression you expect either all concrete `TelemetryInfo` subclasses to be defined by vendors, or at least the all downstream (i.e. all) `Destination` classes to know how to convert them to whatever the Destination wants. In my view, I'd prefer it if a) there was the option to have some (but certainly not all) `TelemetryInfo` subclasses in the upstream LLVM repository, and b) a way to avoid repeating more than is fundamentally necessary to do the serialization. This is a strategy we use successfully in our internal telemetry project.

In my mind, unless we have concrete instances of at least `TelemetryInfo` in the upstream codebase, and therefore the ability for upstream tools to dispatch telemetry with no more than a vendor-specified Destination (specified in just one place), the whole upstream framework becomes pointless, because downstream vendors still end up with the same level of patches needed to actually make use of the framework.

Regarding the details, I'd expect the theoretical `JsonSerializer` class to make use of the `llvm::json` interface, so converting the input types passed by the TelemetryInfo subclass `serialize` method into the appropriate Json types (aside: this is probably generic enough that it could live in the core LLVM code). Similarly, a `ProtobufSerializer` would do something equivalent. The keys are intended to be string literals typically that the `TelemetryInfo` subclasses specify in their `serialize` method - I'm not quite sure what you mean by the "right key string" in this context, as I'm not familiar with protobuf; it would help if you could elaborate some more what the issue here is.

Reply:
> I'm wondering if we've got slightly different visions of things and it's driving the way we're looking at the design differently. 

I think we have very similar ideas of how this ought to work, but there's some miscommunication probably. (ie., , your second paragraph describes pretty much how it is implemented, at least for LLDB's telemetry)


> Reading between the lines of some of the comments and responses you've made, I get the impression you expect either all concrete `TelemetryInfo` subclasses to be defined by vendors, or at least the all downstream (i.e. all) `Destination` classes to know how to convert them to whatever the Destination wants. 

 I want to point out that we should not conflate vendor-specific  code  with tool-specific code.
While it is true that I do not propose any concrete implementation here in `llvm::telemetry`, there *is* a concrete implementation in `lldb::telemetry`. 

So - there are two levels of (potential) overrides here:
 - One: by a tool that wants to use telemetry (eg., LLDB, clang, etc). 
 - Two: by different vendors (for the same tool).

I would expect, the bulk of code implementing subclasses of `TelemetryInfo` to be *upstream* in the individual tool's directory. That is to say, all of LLDB's subclasses of `TelemetryInfo` would live in `lldb/telemetry/...` (but not here in `llvm/telemetry`, because, as pointed out before, different tools care about different events and concepts).

Then, different vendors (for the same tool) will provide the concrete implementation of `Destination`. In this sense, I believe the `Destination` class does exactly what you envision the `Serializer` class would (ie., it takes a `TelemetryInfo` object and picks some or all of the data from that object and forward them elsewhere).

In other words, the only vendor-defined pieces are the concrete implementation of `Destination`
If you would like to have some share utils (eg., how to take a `TelemetryInfo` and turns it into a json object, you could still do that by putting the code in the tool's upstream directory, then your vendor-specific code can call into that)

Reply:
P.S: Concretely, it could look something like this:


```
// In upstream  LLDB: 
llvm::Error lldb::telemetry::Manager::atStartup( TelemetryInfo* info) {
        // Dispatch the `info` to a bunch of pre-registered destinations:
       for (Destination* dest : m_destinations)
             dest->emitEntry(info);
}

llvm::Error lldb::telemetry::Manager::atMainExecutableLoad( MainExecutableInfo* info) {
        // Dispatch the `info` to a bunch of pre-registered destinations:
       for (Destination* dest : m_destinations)
             dest->emitEntry(info);
}

llvm::Error lldb::telemetry::Manager::atDebugCommandExecution( MainExecutableInfo* info) {
        // forward the `info` to a bunch of pre-registered destinations:
       for (Destination* dest : m_destinations)
             dest->emitEntry(info);
}

// In some hypothetical vendor code

llvm::Error lldb::google::InternalStorageDestination::emitEntry(TelemetryInfo* info)  override{
     switch(info->getKind()) {
              // ... dispatch to the corresponding type handler
    }
}

llvm::Error lldb::google::InternalStorageDestination::emitMainExecEntry(MainExecutableInfo* info) {

     google::MainExecutableProto p = make_proto<>();
    // pick out the fields we want to record from the argument and copy them into the proto.

    // send the proto to final storage:
   sendToInternalStorage(std::move(p));
}

```

Reply:
> Regarding the details, I'd expect the theoretical `JsonSerializer` class to make use of the `llvm::json` interface, so converting the input types passed by the TelemetryInfo subclass `serialize` method into the appropriate Json types (aside: this is probably generic enough that it could live in the core LLVM code). Similarly, a `ProtobufSerializer` would do something equivalent. The keys are intended to be string literals typically that the `TelemetryInfo` subclasses specify in their `serialize` method - I'm not quite sure what you mean by the "right key string" in this context, as I'm not familiar with protobuf; it would help if you could elaborate some more what the issue here is.

Let me try to elaborate on this (I was discussing this with Vy last week). Protobufs are more strongly typed than json (you can basically think of them as structs which know how to serialize themselves to a byte stream). The natural way to work with them is to work with named accessors (`my_proto.set_field1(value1); my_proto.set_field2(value2)`). While they have a reflection API which allows accessing the fields using strings (which means a protobuf serializer could implement a virtual `write(StringRef key, ...)` function, it's a rather long-winded way of using them (you convert the field to a string in the TelemetryInfo object, and then immediately undo that in the reflection API. Since vendor "knows" about all of the telemetry types it wants to collect (*), it could do it more directly with `my_proto.set_field1(info.field1)`.

The second source of confusion (now I'm moving on to the "right key string" part) is that this `serializer.write` API (at least in the way I/we understand it) is only suitable for implementing simple dictionary types (`map<string, BasicType>`). It's not very suitable for more complex value types like lists or (sub)dictionaries, as you'd have to add a new (virtual) function for each new data type. (This can be avoided by creating some sort of a polymorphic value type that can hold arbitrary structured data, but at that point, we're sort of reinventing json::Value.)

(*) we assume that to be the case since the vendor needs to decide (based on its use case, data collection policy, etc.) whether a particular field can be collected or not

Reply:
Thanks for the explanations - I now see what you're thinking.

> I want to point out that we should not conflate vendor-specific code with tool-specific code.

This is a fair point. The key thing this means is that code in the top-level framework that is designed to be common needs to be able to work with subclasses defined in tool-level and vendor code. In particular, I think it's reasonable to expect there to be TelemetryInfo subclasses defined in common LLVM code (so for example a `DurationEvent` might be something desired by LLD, clang, llvm-objdump etc), in tools (for events specific to particular tools like LLDB) and in vendor code (for events that downstream vendors want which might not be applicable upstream).

I think it's also reasonable to assume that more event types will be written over time (and existing events will be extended). What I think we want to avoid though is for the need for every `Destination` (or some other class) to need to be extended every time a new event type is implemented. Indeed, it's going to be near impossible for downstream vendors to keep tabs on additions to upstream events in the current structure, I think, so they won't know whether there's a need to review new fields etc, with the current design. I guess the flipside of this is that if some vendors need to review fields for legal reasons, we need to make it clear to them that something has changed that needs review. Mind you, I think there'd still be cases where vendors need to review things without any changes in TelemetryInfo classes at all: imagine the case where someone changes the thing used to populate a field upstream, e.g. to use full paths instead of filenames - for some vendors, having the full path may not be acceptable.

Sorry, that last paragraph is a bit rambly, but I'm leaving it as-is, so that you can see some of the thoughts going through my head. 

Ultimately, I think my point is that the advantage of having something like `TelemetryInfo::serialize` is that the `Destination` classes don't necessarily all need modifying every time a new event field or type is added. I'm open to other ways of how this might be achieved.

I'll acknowledge that something strongly typed like protobuf doesn't work well with the `serialize` method approach, because of the extra indirection. However, you could have `ProtobufDestination` not use the `serialize` approach at all and dispatch by switching based on type, as described. If we allowed both options, it would admittedly mean the `serialize` method wouldn't always be used, but it would allow some `Destination` types to make use of it.

> The second source of confusion (now I'm moving on to the "right key string" part) is that this `serializer.write` API (at least in the way I/we understand it) is only suitable for implementing simple dictionary types (`map<string, BasicType>`). It's not very suitable for more complex value types like lists or (sub)dictionaries, as you'd have to add a new (virtual) function for each new data type. (This can be avoided by creating some sort of a polymorphic value type that can hold arbitrary structured data, but at that point, we're sort of reinventing json::Value.)

FWIW, in our downstream implementation, our `Serializer` base class declares `startArray`/`endArray`/`startObject`/`endObject` methods that our concrete serializers then implement for their different types. There's also a concrete method defined in the base class for serializing iterable types like arrays and vectors, and another for maps (the latter currently only supports `std::unordered_map`, but there's no reason it couldn't be extended to all map-like objects). Clearly for other complex object types, you'd need to define how to serialize that object type, though I've only rarely needed to resort to this in our downstream code.

Reply:
Ok, I've thought about this for awhile, I think we can go with James's `Serializer`  proposal.
(To handle the protobuf case, we can, as pointed, simply not use the Serializer).

Reply:
(NOTE: to be implemented ... this is not resolved yet)